### ISSUE TITLE
Batchnorm statistics were updated using .train() mode, may cause unconsistency with student

### ISSUE THREAD

**Issue by hackerekcah** [hackerekcah/asc_mean_teacher#1]
https://github.com/hackerekcah/asc_mean_teacher/blob/055d0dd514210629f23c1c84ef2246d16378b5bb/role.py#L69-L72

**Comment by hackerekcah** [comment]
use .eval() should not influence too much, just disabled dropout during eval

**Comment by hackerekcah** [comment]
## Batch Norm layer: variables list as follow
```
bn1.weight torch.Size([32])
bn1.bias torch.Size([32])
bn1.running_mean torch.Size([32])
bn1.running_var torch.Size([32])
bn1.num_batches_tracked torch.Size([])
```
## Learnable parameters list:
```
bn1.weight
bn1.bias
```

EMA update wont update teacher's statistics, thus teacher and student have different mean and var,
set .eval() would not update mean and var, thus would be wrong.
During test, statistics seen by student and teacher is different.

## Debuging Info

for key, val in self.model.state_dict().items():
    print(key, val.shape)

conv1.weight torch.Size([32, 1, 7, 7])
conv1.bias torch.Size([32])
bn1.weight torch.Size([32])
bn1.bias torch.Size([32])
bn1.running_mean torch.Size([32])
bn1.running_var torch.Size([32])
bn1.num_batches_tracked torch.Size([])
conv2.weight torch.Size([64, 32, 7, 7])
conv2.bias torch.Size([64])
bn2.weight torch.Size([64])
bn2.bias torch.Size([64])
bn2.running_mean torch.Size([64])
bn2.running_var torch.Size([64])
bn2.num_batches_tracked torch.Size([])
fc1.weight torch.Size([100, 128])
fc1.bias torch.Size([100])
fc2.weight torch.Size([10, 100])
fc2.bias torch.Size([10])
```

```
for p in list(self.model.named_parameters()):
    print(p[0])

conv1.weight
conv1.bias
bn1.weight
bn1.bias
conv2.weight
conv2.bias
bn2.weight
bn2.bias
fc1.weight
fc1.bias
fc2.weight
fc2.bias
```
```
